src/portable/st/synopsys/dcd_synopsys.c

```diff
@@ -576,6 +576,12 @@ static void handle_rxflvl_ints(USB_OTG_OUTEndpointTypeDef * out_ep) {
 
         // Increment pointer to xfer data
         xfer->buffer += bcnt;
+
+        // Truncate transfer length in case of short packet
+        if(bcnt < xfer->max_size){
+          xfer->total_len -= (out_ep[epnum].DOEPTSIZ & USB_OTG_DOEPTSIZ_XFRSIZ_Msk) >> USB_OTG_DOEPTSIZ_XFRSIZ_Pos;
+          if(epnum == 0) xfer->total_len -= ep0_pending[TUSB_DIR_OUT];
```

Comment:
ep0_pending[TUSB_DIR_OUT] should be zero as well I think to end the whole transfer

Reply:
I think `ep0_pending` in this case can have a different value than zero. It contains the bytes that do not fit into `DOEPTSIZ0`. Therefore `XFRSIZ` of `DOEPTSIZn` of different epnums should be equal to `XFRSIZ` of `DOEPTSIZ0` of EP0 **+** `ep0_pending`.

e.g. we request 512 bytes to be received at EP0. We schedule 64 bytes and set ep0_pending to 448. If the first scheduled packet has short length 10 the transfer will be already finished. Then `DOEPTSIZ` will contain 54 bytes only and `ep0_pending` has to be used to get the real remaining bytes of 502.

Reply:
Ahh sorry I missunderstood you. Yes we should **set** ep0_pending to zero here. 🙄